### ISSUE TITLE
linear region with color array

### ISSUE THREAD

**Issue by shlomok613** [vispy/vispy#1697]
We wish to use vispy generate a real time Sonogram like display.

We need to push a line the the top of the visual and move the others lines down.

Currently we are using a list of line visuals to display our data the update rate is really slow for a large number (> 20) lines as each visual needs to redrawn.

The Linear Region would be ideal for our needs as we can easily rotate the colors as add the new color to the top. Performance is great for large numbers of lines. 
Except that the linear region it only accepts a single color for each line while we need to display an array of colors for each line.

What would be involved in modifying the Linear Region to accept an array of colors? 

Thanks for your time.

**Comment by djhoese** [comment]
As you've found, using many visuals can be very slow as the GPU has to draw each one separately, one at a time. Searching for a solution where one visual draws multiple lines will be better for performance.

I've never personally used the LinearRegionVisual, but based on the documentation it already supports multiple colors: https://github.com/vispy/vispy/blob/26118b87af9af5ea60bde1df1d0594e7b41e4a68/vispy/visuals/linear_region.py#L45

If something isn't working, you could compare the linear region code to the changes in [this pull request](https://github.com/vispy/vispy/pull/1558/files#diff-0ead966286007acd9fb58a3dfafcb0c2) where multiple colors were added to the TextVisual.

**Comment by shlomok613** [comment]
LinearRegion currently plots an list of single rgba values one for each each bar.

I need to able to plot a list of of lists of rgba colors one list for each bar so I display an array of color maps.

How do I modify the gloo code to plot an array of vectors one of each bar instead of an array of scalars?

Any ideas?

Shlomo

**Comment by djhoese** [comment]
Do you have an image of what you are trying to do?

Things on screen can only have one color at a time. You either change the color for a particular vertex/fragment (by reassigning the color on the visual's property) or you could load all possible colors in to the GPU and use an index array to say the GLSL equivalent of "this vertex should be `colors[color_idx]`".

If the LinearRegion, and the `linear_region.py` example, do allow you to show colors in the way you need, you'll need to describe what you want in a way that could be described on the GPU. Let me know if you have an example of what you want. Here is the `linear_region.py` example:

![image](https://user-images.githubusercontent.com/1828519/63651688-2594d280-c71d-11e9-83fc-b2a539b82044.png)

**Comment by shlomok613** [comment]
We are developing a position sensor with 3 optical sensors C P L. We to be able display the raw sensor data as 3 graphs and a history expressed as a line strip for each sensor set captured. We call this history the waterfall plot. This first line strip is pushed in to the top of the of the history graph and the last one is thrown away. Each line strip is an array of about 3000 colors (could reduce to 200) generated from the raw sensor data. See the attached image
![waterfall](https://user-images.githubusercontent.com/52164257/63669006-81c12a80-c7e1-11e9-8063-effa28e9322a.PNG)


We currently do the waterfall plot using 60 scene.Line plots which is very slow. If we pass an array of colors to the linear region we simulate what we want to do with a single color. We have modified linear.py to add a yellow line on top of a list of plots and rotate those plot downward See the attached code and and graphic.

[linear.txt](https://github.com/vispy/vispy/files/3539693/linear.txt)

![linear](https://user-images.githubusercontent.com/52164257/63670173-a5d23b00-c7e4-11e9-8e54-9a4450755f3d.PNG)



What we want to able to do is to modify LinearRegion to accept an list of list of colors in place of the single colors currently supported.

Thanks for you help,

Shlomo

**Comment by kmuehlbauer** [comment]
@shlomok613 Maybe I'm missing something, but you might use three `ImageVisual` for the waterfall.

**Comment by djhoese** [comment]
I agree with @kmuehlbauer that I don't think a LinearRegion is the right visual for this use case. The LinearRegion is really meant (as far as I can tell from looking at the code) to only have single vertical or horizontal regions. To make it do what you want you'd be essentially make M numbers of lines with N numbers of vertexes (one for each color transition on the line). This is essentially an M x N image. Although you are adding individual lines at a time, you can think of them as rows of an image.

You may still have to do some customization of the ImageVisual to get really good performance. Ideas:

1. Things like what's implemented in #1607 which I hope to merge before the next release after some more review. This would let you update portions of the image (a single line) and only send that portion of the image array to the GPU. However, this gets difficult since you need to "roll" the image data.
2. You could modify the ImageVisual to send raw data to the GPU instead of the data adjusted for the color limits (`clim`). I've done something similar to this in my [own project](https://github.com/ssec/sift/blob/master/uwsift/view/visuals.py#L248) but it may be hard to follow. I hope to add this to vispy in the future, but right now it isn't the default because it is not OpenGL ES compatible. The color assignment can be done in the GPU (with a colormap) instead of having to generate an RGB in the CPU and sending it to the GPU (what you'd be doing with the line region).
3. In addition to 1, you could make the ImageVisual have two texture "tiles" (quads). As you update data in the Texture2D with new line data, instead of adding a new line and removing the old line you could *overwrite* the old line with the new line. Then you define one quad to be the top half of the image data array and the second quad is the bottom half. This would avoid the "roll" of the image data and make it so you could only send the new line (needs #1607 to work completely).

All of these techniques are to avoid the amount of data that gets sent to the GPU and avoid doing a lot of operations on that data before sending it. One complication I can envision is if your line pixels (column width) are not equally spaced since the ImageVisual by default will assume the pixels are equally spaced.

</braindump>

**Comment by shlomok613** [comment]
Thanks for the detailed response.
Please could you send me an example of how to use the TiledGeolocatedImage so I can get started.

Please let know when you have merged #1607.

**Comment by shlomok613** [comment]
Why  three ImageVisual for the waterfall?

**Comment by kmuehlbauer** [comment]
> Why three ImageVisual for the waterfall?

In your example image there are three sensors, right? So I would add one waterfall for each sensor.

**Comment by shlomok613** [comment]
I get data at about 30 fps. 

I am still not too clear about quads. Do you mean to overwrite the top line and somehow remap the data so that all the other lines magically move down? How would this work?

I tried with a regular image visual and I found that it was too slow. I will try again with  #1607  and smaller color array and let you know how it goes.

**Comment by shlomok613** [comment]
Thanks for taking the time to explain things so clearly.

How do we tell the image visual to draw the newest line at the top of the plot and the oldest line at the bottom? 

Are you suggesting two image visuals one for the top quad and one for the bottom quad?

**Comment by shlomok613** [comment]
Wow, @os-gabe ... I hadn't dared try to do this with 3D... your visuals look great.

It seems that this approach would be incompatible with the "Two Quad" approach proposed by @djhoese (which does a 2D scroll, not a 3D one.
If I understand your code correctly, the data is still being "rolled" by the deque (possibly switch to doing it in Numpy) which means the entire data-set has to be squirted to the GPU every time it is rolled... which might not bode well for performance if we move towards data-sets that contain a full-screen's worth of data points.
Am I missing something here? Would @djhoese's patch  #1607 help here?

@djhoese, I would like to experiment with writing a custom ImageVisual, but as someone new to Vispy, am struggling both with how to get textured line data into the quads and supect I'll also find it confusing to get the two adjacent quads to appear on screen. If you have code that does anything somewhat similar that you can share here, I'd be much obliged to have it as a starting point.


Thanks!
Shlomo

**Comment by djhoese** [comment]
I have not tried @os-gabe's solution to see what it produces, but it all depends on what you want it to look like. The ImageVisual and my two quad solution should work fine in 3D but it would essentially be a 2D plane in a 3D space.

Sorry @shlomok613, I don't have time to put together an example of the quad stuff. @os-gabe's solution may be a good start if you like the way it looks. It should perform ok as-is, but it could benefit from #1607 in the future (I think).

**Comment by shlomok613** [comment]
@djhoese, I am about to start work on my custom ImageVisual WaterfallImageVisual. I have a few questions:

How I deal with mapping the two quads? 

Do I use two image visuals and change vertices for each each time I add a new row? 

Or do I have one image visual in which case how do I handle wraparound from the bottom to the top?

Any other advice that you can give me would be very welcome.

Shlomo

**Comment by shlomok613** [comment]
here is my first attempt with a wraparound. It doesn't work

class WaterfallImageVisual(ImageVisual):
    def __init__(self,parent,max_tile_width,max_tiles):
        self._max_tile_width=max_tile_width
        self._max_tiles=max_tiles
        self._row_index=0
        grid=(self._max_tile_width,self,max_tiles)
        self._data = np.zeros((self._max_tiles,self._max_tile_width),dtype=np.float32)
        cmap = Colormap(['b', 'y', 'w'])
        super().__init__(data=self._data,parent=parent,grid=grid,cmap=cmap,clim='auto',method='subdivide')

    def add_row(self,row):
        assert(len(row.shape)== 1 and row.shape[0] == self._max_tile_width)
        if self._row_index >= self._max_tiles:
            #wraparound for now
            self._row_index =  0
        self._data[self._row_index] = row
        self._row_index += 1
        self._need_texture_upload = True
        self._need_vertex_update = True
        self.update()

Please could you help?

**Comment by djhoese** [comment]
You have one image visual for each waterfall (you had three waterfalls in your original image). Take a look at how the builtin ImageVisual builds the vertex coordinates and texture coordinates [here](https://github.com/vispy/vispy/blob/master/vispy/visuals/image.py#L341-L366).

In this method it is making one "starter" quad (`quad`) then repeats it for the number of "tiles" the image is split in to (see `grid` keyword argument). Your approach is different in that you don't want to make "tiles" of quads, you only want two. Also note how this method is creating vertex coordinates (where on the screen the quads go) and texture coordinates (where the image data is in the texture object).

You can format your code on github comments by putting three backticks (`) before and after the block of code you want to format. The backticks must be on their own line. It is a little hard to read your code, but what are you setting your max tile and tile width to? In my mind `grid` could almost be hard-coded to `(2, 1)`, 2 rows, 1 column, for each chunk of the image. You then need to update the `_build_vertex_data` to build the vertex and texture coordinates properly.

**Comment by shlomok613** [comment]
Thanks for the quick reply

For testing purposes the number of tiles is set to 10 and tile width is set to 200.

I am having trouble understanding the _build_vertex_data method. Please could you explain it to me and tell me what needs to be changed.

How do I update the vertex and texture coordinates when I add a new tile?

**Comment by shlomok613** [comment]
In the final application we plan to use one image visual and concatenate the 3 plots as we get all the data at one time. 

We are doing this currently for the line visual and it is using a tiny amount of the CPU

we intend to use 100 lines of about 800 pixels.

**Comment by djhoese** [comment]
What does 10 tiles mean to you and your application? Why do you need tiles at all? Each tile is a quad in OpenGL-land so you should really only need two per waterfall.

Each tile is a quad. The `quad` in `_build_vertex_data` is made up of two triangles (3 vertices per triangle). This quad is then repeated to generate a quad for each tile. The texture coordinate quads have coordinates between 0 and 1. The vertex coordinates are whatever coordinate system you want them to have, by default ImageVisual is using a coordinate system that's the same size as the image (a distance of 1 in vertex coordinate space == 1 pixel on the image).

I would suggest running the `_build_vertex_data` code outside of vispy by itself to see what it is actually producing. The only inputs are width, height, and the number of tiles in each of those dimensions (`self._grid`). Maybe draw out where each of the coordinates on a piece of paper. Otherwise, it may be good to google how texture coordinates work.

**Comment by shlomok613** [comment]
I am working on implementing the WaterfallImageVisual. I have a few questions:

I am sub classing the ImageVisual.

I now under how to make the two quads and twiddle their vertices

How do I display the two quads? Do I use a FBO as in https://github.com/vispy/vispy/blob/master/examples/basics/gloo/hello_fbo.py ?

Any other implementation tips would be most welcome.

Shlomo

**Comment by djhoese** [comment]
You are "making" quads by creating 6 vertex coordinates and 6 texture coordinates for each quad (2 triangles == 1 quad). You won't need to add any new low-level GL stuff to the existing ImageVisual. You're defining two boxes. The GPU takes your texture coordinates and says "I'm going to take this chunk of the texture (this rectangle/quad) and place it on the canvas at these vertex coordinates" then it moves on to the next quad in the coordinates.

Does that help at all?

**Comment by shlomok613** [comment]
I present a simple demo with have two quads and two textures shown below.
Do I need the two sets of shaders? I am using them so that I can map a texture to a particular quad.

How would I do this with only one set of shaders?

Do I need to resize the quads height. If so how?

I am as you see just learning open GL an very much appreciate your help.

Shlomo

```"""
Simple example demonstrating showing two quad.
"""

from vispy import gloo
from vispy import app
import numpy as np

image_bottom = np.zeros((100, 100, 3), 'float32')
image_bottom[:50, :, 0] = 1.0
image_bottom[:, :50, 1] = 1.0
image_bottom[50:, 50:, 2] = 1.0

image_top = np.zeros((100, 100, 3), 'float32')
image_top[:50, :50, 0] = 1.0
image_top[:, :50, 1] = 1.0
image_top[50:, 0:, 2] = 1.0
# Create vertices
bottom_quad = np.array([[-1.0, -1.0, 0.0], [+1.0, -1.0, 0.0],
                      [-1.0, +0.0, 0.0], [+1.0, +0.0, 0.0]], np.float32) #bottom quad

top_quad    = np.array([[-1.0, 0.0, 0.0], [1.0, 0.0, 0.0],
                       [-1.0, 1.0, 0.0],   [1.0, 1.0, 0.0,]], np.float32) #top quad

texture_quad = np.array([[0.0, 0.0], [0.0, 1.0],[1.0, 0.0], [1.0, 1.0]],np.float32)

VERT_TOP = """ // simple vertex shader
attribute vec3 a_position;
attribute vec2 a_texcoord;

void main (void) {
    gl_TexCoord[0] = vec4(a_texcoord.x, a_texcoord.y, 0.0, 0.0);
    gl_Position = vec4(a_position, 1.0);
}
"""

FRAG_TOP = """ // simple fragment shader
uniform sampler2D a_texture;
void main()
{
    gl_FragColor = texture2D(a_texture, gl_TexCoord[0].st);
}
"""

VERT_BOTTOM = """ // simple vertex shader
attribute vec3 a_position;
attribute vec2 a_texcoord;

void main (void) {
    gl_TexCoord[0] = vec4(a_texcoord.x, a_texcoord.y, 0.0, 0.0);
    gl_Position = vec4(a_position, 1.0);
}
"""

FRAG_BOTTOM = """ // simple fragment shader
uniform sampler2D a_texture;
void main()
{
    gl_FragColor = texture2D(a_texture, gl_TexCoord[0].st);
}
"""

class Canvas(app.Canvas):

    def __init__(self):
        app.Canvas.__init__(self, keys='interactive')

        # Create program
        self._program_top = gloo.Program(VERT_TOP, FRAG_TOP)
        self._program_bottom = gloo.Program(VERT_BOTTOM, FRAG_BOTTOM)

        self._program_top['a_position'] = gloo.VertexBuffer(top_quad)
        self._program_top['a_texcoord'] = texture_quad
        self._program_top['a_texture'] = gloo.Texture2D(image_top)

        self._program_bottom['a_position'] = gloo.VertexBuffer(bottom_quad)
        self._program_bottom['a_texcoord'] = texture_quad
        self._program_bottom['a_texture'] = gloo.Texture2D(image_bottom)

        gloo.set_clear_color('white')

        self.show()

    def on_resize(self, event):
        width, height = event.physical_size
        gloo.set_viewport(0, 0, width, height)

    def on_draw(self, event):
        gloo.clear()
        self._program_top.draw('triangle_strip')
        self._program_bottom.draw('triangle_strip')


if __name__ == '__main__':
    c = Canvas()
    app.run()
```

**Comment by djhoese** [comment]
Why aren't you subclassing the ImageVisual? All the code is already written there.

To have more than one quad in a coordinate array, keep adding points to the array. You *should* have 6 coordinates (2 triangles) for each quad so to add another quad you would take your `(6, 3)` shape array and add 6 more points to it to make it `(12, 3)`. The GPU is told to draw triangles so it will take every 3 points as a new triangle by default (at least in the ImageVisual).

**Comment by shlomok613** [comment]
Here is my implementation of  _build_vertex_data.
Have I got the shape array right?
What do I need to multiply by fraction in order to get the waterfall effect to work?
Have I missed anything?
```
 def _build_vertex_data(self):
        """Rebuild the vertex buffers used for rendering the image when using
        the subdivide method.
        """
        grid = self._grid
        w = 1.0 / grid[1]
        h = 1.0 / grid[0]

        index = self._index
        max_rows = self._height
        fraction = index / max_rows
        
        quad = np.array([
                         [0, 0, 0], [w, 0, 0], [w, h/2, 0],
                         [0, 0, 0], [w, h/2, 0], [0, h/2, 0],
                         [0, h/2, 0], [w, h/2, 0], [w, h, 0],
                         [0, h/2, 0], [0, h, 0], [w, h, 0],
                         ],
                        dtype=np.float32)
        quads = np.empty((grid[1], grid[0], 12, 3), dtype=np.float32)
        quads[:] = quad

        mgrid = np.mgrid[0.:grid[1], 0.:grid[0]].transpose(1, 2, 0)
        mgrid = mgrid[:, :, np.newaxis, :]
        mgrid[..., 0] *= w
        mgrid[..., 1] *= h

        quads[..., :2] += mgrid
        tex_coords = quads.reshape(grid[1]*grid[0]*12, 3)
        tex_coords = np.ascontiguousarray(tex_coords[:, :2])
        vertices = tex_coords * self.size

        self._subdiv_position.set_data(vertices.astype('float32'))
        self._subdiv_texcoord.set_data(tex_coords.astype('float32'))
```

**Comment by djhoese** [comment]
This is all guessing and I would encourage you to print out the vertices and tex_coords arrays to see if they make sense. I also think you can cut out some of the complexity if you cut out anything that depends on the `self._grid` stuff. Since you don't want to make any mini-quads I think. If I recall correctly this is only useful if you'll be zooming in on the waterfall and want to ease what the GPU has to draw by only drawing part of the waterfall (a subset), but I think you said you'll be viewing the whole thing the whole time. Even if not, it isn't that much data for the GPU.

As for the fraction, I think where you have `h/2` in your `quad` you could have `fraction * quad`. Right?

**Comment by shlomok613** [comment]
why not not fraction * h/2 ?

**Comment by djhoese** [comment]
Actually I think you could just use `fraction` since you want a number between 0 and 1 for the `quad` portion of the code which will later get multiplied by the height. Multiplying by the height would give you the "row" of the image (same as index in this case) which you need for the vertex coordinate, but you also need the actual 0 to 1 fraction for the texture coordinate. So vertex coordinates are from 0 to width/height and texture coordinates are from 0 to 1.

**Comment by shlomok613** [comment]
I am still not getting the waterfall effect. What I am doing wrong
```
class WaterfallImageVisual(scene.visuals.Image):
    def __init__(self,parent,height,width):
        self.unfreeze()
        self._height = height
        self._width = width
        self._data = np.zeros((height, width))
        cm = Colormap(['w', 'r', 'g', 'b'])
        self._index = 0
        super().__init__(data=self._data,parent=parent,cmap=cm,clim=[0,1])

    def _build_vertex_data(self):
        """Rebuild the vertex buffers used for rendering the image when using
        the subdivide method.
        """

        index = self._index
        max_rows = self._height
        fraction = index / max_rows

        quad = np.array([
                         [0, 0, 0], [1, 0, 0], [1, fraction, 0],
                         [0, 0, 0], [1, fraction, 0], [0, fraction, 0],
                         [0, fraction, 0], [1, fraction, 0], [1, 1, 0],
                         [0, fraction, 0], [0, 1, 0], [1, 1, 0],
                         ],
                        dtype=np.float32)
        quads = np.empty((1,1, 12, 3), dtype=np.float32)
        quads[:] = quad
        

        mgrid = np.mgrid[0.:1, 0.:1].transpose(1, 2, 0)
        mgrid = mgrid[:, :, np.newaxis, :]

        quads[..., :2] += mgrid
        tex_coords = quads.reshape(12, 3)
        tex_coords = np.ascontiguousarray(tex_coords[:, :2])
        vertices = tex_coords * self.size

        self._subdiv_position.set_data(vertices.astype('float32'))
        self._subdiv_texcoord.set_data(tex_coords.astype('float32'))


    def add_row(self,row):
        # print("add_row index=",self._index)
        self._data [self._index] = row
        self._index = (self._index + 1) % self._height
        self.set_data(self._data)
```

**Comment by djhoese** [comment]
If you look at the ImageVisual's `set_data` method you'll see why. The `set_data` method assumes that if the new data being set is the same shape as the old data then it doesn't need to re-build the vertex data. You'll probably want to override to make your own `set_data` *or* since you are already modifying `self._data` in `add_row` maybe just set the appropriate update flags like `set_data` does:

```
def add_row(self, row):
    ... existing code except for the set_data call ...
    self._need_vertex_update = True
    self._need_texture_upload = True
```

Then the next time the draw is done (see ImageVisual's prepare_draw method) the vertex and texture coordinates (your _build_vertex_data method will be called) and the new texture data will be sent to the GPU.

**Comment by queenish001** [comment]
@shlomok613 Is your query resolved now?

**Comment by djhoese** [comment]
I'm going to close this as we went through a couple options it seems. If this is still an issue that we can actively discuss and work on then we can reopen it. If you have other issues then lets do that in another issue.
